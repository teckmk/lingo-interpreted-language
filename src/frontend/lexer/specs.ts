import { EOL } from "os"

export enum TokenType {
  NumberLiteral = "NUMBER_LITERAL",
  StringLiteral = "STRING_LITERAL",
  Identifier = "IDENTIFIER",
  SingleLineComment = "SINLE_LINE_COMMENT",
  DocComment = "DOC_COMMENT",
  MultiLineComment = "MULTI_LINE_COMMENT",
  WhiteSpace = "WHITE_SPACE",
  Tab = "TAB",
  Indent = "INDENT",
  Dedent = "DEDENT",
  EOF = "END_OF_FILE",
  EOL = "END_OF_LINE",

  Equals = "ASSIGNMENT_OPERATOR", // =
  Comma = "COMMA", // ,
  Dot = "DOT", // .
  Exclamation = "EXLAMATION", // "!"
  Colon = "COLON", //":"
  OpenParen = "OPEN_PAREN", //  (
  CloseParen = "CLOSE_PAREN", // )
  OpenBrace = "OPEN_BRACE", // {
  CloseBrace = "CLOSE_BRACE", // }
  OpenBracket = "OPEN_BRACKET", // [
  CloseBracket = "CLOSE_BRACKET", // ]
  LogicGate = "LOGIC_GATE", // && || and or
  AdditiveOperator = "ADDITIVE_OPERATOR", //  + -
  MulitipicativeOperator = "MULTIPICATIVE_OPERATOR", // / * %
  EqualityOperator = "EQUALITY_OPERATOR", // == !=
  RelationalOperator = "RELATIONAL_OPERATOR", // > < >= <=

  // Reserved
  // KeyWords
  Let = "LET",
  Const = "CONST",
  Final = "FINAL",
  Fn = "FN",
  Return = "RETURN",
  If = "IF",
  Else = "ELSE",
  While = "WHILE",

  // Primitive Types
  NumberType = "NUMBER_TYPE",
  StringType = "STRING_TYPE",
  BooleanType = "BOOL_TYPE",
  DynamicType = "DYNAMIC_TYPE",
}

export type Spec = {
  regex: RegExp
  tokenType: TokenType
}

export const specs: Spec[] = [
  { regex: /^[+-]?([\d]*[.])?[\d]+/, tokenType: TokenType.NumberLiteral },
  { regex: /^"[^"]*"/, tokenType: TokenType.StringLiteral },
  { regex: /^'[^']*'/, tokenType: TokenType.StringLiteral },

  { regex: /^\bvar\b/, tokenType: TokenType.Let },
  { regex: /^\bconst\b/, tokenType: TokenType.Const },
  { regex: /^\bfinal\b/, tokenType: TokenType.Final },
  { regex: /^\bfn\b/, tokenType: TokenType.Fn },
  { regex: /^\breturn\b/, tokenType: TokenType.Return },
  { regex: /^\bif\b/, tokenType: TokenType.If },
  { regex: /^\belse\b/, tokenType: TokenType.Else },
  { regex: /^\bwhile\b/, tokenType: TokenType.While },

  { regex: /^\bnumber\b/, tokenType: TokenType.NumberType },
  { regex: /^\bstring\b/, tokenType: TokenType.StringType },
  { regex: /^\bbool\b/, tokenType: TokenType.BooleanType },
  { regex: /^\bdynamic\b/, tokenType: TokenType.DynamicType },

  { regex: /^[a-zA-Z_][a-zA-Z0-9_]*/, tokenType: TokenType.Identifier },

  { regex: /^\/\/.*/, tokenType: TokenType.SingleLineComment },
  { regex: /^\/\/\/.*/, tokenType: TokenType.DocComment },

  // these are making problem with line numbers, because they are read as single line, no matter how many lines they span on....
  // { regex: /^\/\*[\s\S]*?\*\//, tokenType: TokenType.MultiLineComment },

  // operators
  { regex: /^[+-]/, tokenType: TokenType.AdditiveOperator },
  { regex: /^[*/%]/, tokenType: TokenType.MulitipicativeOperator },
  { regex: /^[><]=?/, tokenType: TokenType.RelationalOperator },
  { regex: /^[=!]=/, tokenType: TokenType.EqualityOperator },

  { regex: /^&&/, tokenType: TokenType.LogicGate },
  { regex: /^\|\|/, tokenType: TokenType.LogicGate },
  { regex: /^\band\b/, tokenType: TokenType.LogicGate },
  { regex: /^\bor\b/, tokenType: TokenType.LogicGate },

  { regex: /^:/, tokenType: TokenType.Colon },
  { regex: /^,/, tokenType: TokenType.Comma },
  { regex: /^\./, tokenType: TokenType.Dot },
  { regex: /^=/, tokenType: TokenType.Equals },
  { regex: /^!/, tokenType: TokenType.Exclamation },
  { regex: /^\(/, tokenType: TokenType.OpenParen },
  { regex: /^\)/, tokenType: TokenType.CloseParen },
  { regex: /^{/, tokenType: TokenType.OpenBrace },
  { regex: /^}/, tokenType: TokenType.CloseBrace },
  { regex: /^\[/, tokenType: TokenType.OpenBracket },
  { regex: /^\]/, tokenType: TokenType.CloseBracket },

  { regex: RegExp(`^${EOL}`), tokenType: TokenType.EOL },

  { regex: /^[^\S\r\n]+/, tokenType: TokenType.WhiteSpace }, // match one or more spaces and tabs
]
