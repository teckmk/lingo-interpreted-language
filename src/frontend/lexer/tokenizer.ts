/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Spec, specs, TokenType } from "./specs"

type Position = {
  line: number
  column: number
}

export type Token = {
  type: TokenType
  value: any
  position: { start: Position; end: Position } // end.column is exclusive
}

export class Tokenizer {
  private _spec: Spec[]
  private _tokens: Token[] = []

  private _cursor = 0
  private _line = 1 // current line
  private _tokenNumber = 0 // current token

  private _code = ""
  private _filename = ""

  constructor(spec: Spec[], filename: string) {
    this._spec = spec
    this._filename = filename
  }

  tokenize(code: string) {
    this._code = code
    this._cursor = 0
    this._line = 1
    this._tokenNumber = 0
    this._tokens = []

    while (!this._isEOF && this._hasNext) this._readToken()

    this._pushToken({
      type: TokenType.EOF,
      value: "EOF",
      position: { start: { line: this._line, column: -1 }, end: { line: this._line, column: -1 } },
    })

    return this._tokens
  }

  private get _isEOF(): boolean {
    if (!this._code) return true

    return this._cursor == this._code.length
  }

  private get _hasNext(): boolean {
    if (!this._code || this._code.length == 0) return false

    return this._cursor < this._code.length
  }

  private _readToken() {
    const code = this._code.slice(this._cursor)
    const startCursor = this._cursor

    for (const { regex, tokenType } of this._spec) {
      let tokenValue = this._matched(regex, code)

      // skip iteration if there's no match
      if (tokenValue == null) continue

      // Get the position of the current cursor within the current line
      // We need to calculate the actual column by finding the start position of the current line
      const lineStartIndex = this._code.lastIndexOf("\n", startCursor - 1) + 1
      // Column is 1-indexed
      const startColumn = startCursor - lineStartIndex + 1

      // Calculate the start position
      const startPos = {
        line: this._line,
        column: startColumn,
      }

      // Calculate the end position
      const endPos = {
        line: this._line,
        column: startColumn + tokenValue.length,
      }

      // Handle newlines separately
      if (tokenType == TokenType.EOL) {
        this._line++
        tokenValue = "\n"
      }

      // Process string literals
      if (tokenType == TokenType.StringLiteral) {
        // Unescape escaped quotes and backslashes.
        // TODO: expand this to support other escape sequences.
        tokenValue = tokenValue.replace(/\\"/g, '"').replace(/\\\\/g, "\\")
        // trim the quotes
        tokenValue = tokenValue.slice(1, -1)
      }

      return this._pushToken({
        type: tokenType,
        value: tokenValue,
        position: {
          start: startPos,
          end: endPos,
        },
      })
    }

    throw new Error(
      `Unexpected token '${code[0]}' at ${this._line}:${this._cursor - this._code.lastIndexOf("\n", this._cursor - 1)} in ${this._filename}`,
    )
  }

  private _matched(regex: RegExp, string: string) {
    const matched = regex.exec(string)

    if (matched == null) return null

    this._cursor += matched[0].length
    // Don't increment _tokenNumber here anymore as we're calculating column differently

    return matched[0]
  }

  private _pushToken(token: Token) {
    this._tokens.push(token)
  }
}

export class IndentMaker {
  private _tokens: Token[] = []

  get tokens() {
    return this._tokens
  }

  markIndents(tokens: Token[]): IndentMaker {
    const indentStack: Stack<number> = new Stack([0]) // initialize with root indent 0
    let afterEOL = false
    let afterColon = false

    for (const token of tokens) {
      // If the token is a colon, set the flag and push the token.
      if (token.type === TokenType.Colon) {
        afterColon = true
        this._tokens.push(token)
        continue
      }

      // If there was not EOL immediately after a colon (and afterEOL is not already true), reset the flag.
      if (afterColon && !afterEOL && token.type !== TokenType.EOL) {
        afterColon = false
      }

      // If the token is an end-of-line, push it and note that the next token starts a new line.
      if (token.type === TokenType.EOL) {
        this._tokens.push(token)
        afterEOL = true
        continue
      }

      // If we’re at the start of a new line…
      if (afterEOL && token.type === TokenType.WhiteSpace) {
        const currentIndent = indentStack.peek()
        const newIndent = token.value.length

        // When indent increases, only allow it if the previous line ended with a colon.
        if (newIndent > currentIndent) {
          if (afterColon) {
            indentStack.push(newIndent)
            this._tokens.push(this._createIndentToken(token))
          }
        }
        // When indent decreases, insert one or more dedent tokens.
        else if (newIndent < currentIndent) {
          while (indentStack.length > 1 && indentStack.peek() > newIndent) {
            this._tokens.push(this._createDedentToken(token))
            indentStack.pop()
          }
          if (indentStack.peek() !== newIndent) {
            throw new Error(
              `IndentationError: Invalid indentation at ${token.position.start.line}:${token.position.start.column}`,
            )
          }
        }
        // Reset the flags after processing the whitespace.
        afterEOL = false
        afterColon = false
        // Do not push the whitespace token itself.
        continue
      }

      // If we're at the start of a new line but the line does not start with whitespace…
      if (afterEOL) {
        const currentIndent = indentStack.peek()
        const newIndent = 0
        if (newIndent < currentIndent) {
          while (indentStack.length > 1 && indentStack.peek() > newIndent) {
            this._tokens.push(this._createDedentToken(token))
            indentStack.pop()
          }
          if (indentStack.peek() !== newIndent) {
            throw new Error(
              `Invalid indentation at ${token.position.start.line}:${token.position.start.column}`,
            )
          }
        }
        afterEOL = false
        afterColon = false
        this._tokens.push(token)
        continue
      }

      // For all other tokens, just push them.
      this._tokens.push(token)
    }

    // At the end, add any necessary dedent tokens for any remaining indents.
    while (indentStack.length > 1) {
      indentStack.pop()
      insertAtIndex(
        this._tokens,
        this._tokens.length - 1,
        this._createDedentToken(this._tokens[this._tokens.length - 1]),
      )
    }

    return this
  }

  removeUnwantedTokens() {
    this._tokens = this._tokens.filter(({ type }, index) => {
      const skippable =
        type === TokenType.WhiteSpace ||
        type === TokenType.SingleLineComment ||
        type === TokenType.EOL

      const next = this._tokens[index + 1]
      const nextOfNext = this._tokens[index + 2]
      const indentColon =
        type === TokenType.Colon &&
        next?.type === TokenType.EOL &&
        nextOfNext?.type === TokenType.Indent
      return !skippable && !indentColon
    })
    return this
  }

  fixColumnNumbers() {
    // let currentColumn = 1
    // let currentLine = 1
    // const tokens = []

    // for (const token of this._tokens) {
    //   if (token.type === TokenType.EOL) {
    //     currentLine++
    //     currentColumn = 1
    //     continue
    //   }

    //   if (token.type === TokenType.Indent || token.type === TokenType.Dedent) {
    //     currentColumn = 1
    //   }

    //   const newToken = {
    //     ...token,
    //     column: currentColumn++,
    //     line: currentLine,
    //   }

    //   // the EOF token should have the column number of -1
    //   if (newToken.type === TokenType.EOF) {
    //     newToken.column = -1
    //   }

    //   tokens.push(newToken)
    // }

    // this._tokens = tokens
    return this
  }

  private _createIndentToken(baseToken: Token): Token {
    return {
      type: TokenType.Indent,
      value: "indent",
      position: {
        start: { line: baseToken.position.end.line, column: 1 },
        end: { line: baseToken.position.end.line, column: baseToken.value.length },
      },
    }
  }

  private _createDedentToken(baseToken: Token): Token {
    return {
      type: TokenType.Dedent,
      value: "dedent",
      position: {
        start: { line: baseToken.position.end.line, column: 1 },
        end: { line: baseToken.position.end.line, column: baseToken.value.length },
      },
    }
  }
}

function insertAtIndex<T>(array: T[], index: number, element: T): T[] {
  if (index < 0 || index > array.length) {
    throw new Error("Index is out of range")
  }

  // Use splice to insert the element at the specified index
  array.splice(index, 0, element)

  return array
}

// Create stack class
class Stack<T> {
  private stack: T[] = []

  constructor(initialItems: T[] = []) {
    this.stack = initialItems
  }

  get length() {
    return this.stack.length
  }

  peek(): T {
    return this.stack[this.stack.length - 1]
  }

  push(item: T) {
    this.stack.push(item)
  }

  pop(): T | undefined {
    return this.stack.pop()
  }
}

export function tokenize(filename: string, code: string): Token[] {
  const tokens = new Tokenizer(specs, filename).tokenize(code)

  return new IndentMaker().markIndents(tokens).removeUnwantedTokens().fixColumnNumbers().tokens
}
