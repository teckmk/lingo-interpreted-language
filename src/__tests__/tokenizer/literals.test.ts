import { specs } from "../../frontend/lexer/specs";
import { Tokenizer } from "../../frontend/lexer/tokenizer";

describe("Tokenizer - Literals", () => {
  const tokenizer = new Tokenizer(specs, "test");

  const testCases = [
    { name: "integers", code: "10" },
    { name: "positive integers", code: "+10" },
    { name: "negative integers", code: "-10" },
    { name: "floats", code: "10.5" },
    { name: "positive floats", code: "+10.5" },
    { name: "negative floats", code: "-10.5" },
    { name: "strings", code: '"hello world"' },
    { name: "strings with escaped characters", code: '"hello "world""' },
  ];

  testCases.forEach(({ name, code }) => {
    it(`should tokenize ${name}`, () => {
      expect(tokenizer.tokenize(code)).toMatchSnapshot();
    });
  });
});
