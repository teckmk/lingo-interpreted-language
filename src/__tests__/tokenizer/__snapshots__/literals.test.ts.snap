// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer - Literals should tokenize floats 1`] = `
[
  {
    "position": {
      "end": {
        "column": 5,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10.5",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize integers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize negative floats 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "ADDITIVE_OPERATOR",
    "value": "-",
  },
  {
    "position": {
      "end": {
        "column": 6,
        "line": 1,
      },
      "start": {
        "column": 2,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10.5",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize negative integers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "ADDITIVE_OPERATOR",
    "value": "-",
  },
  {
    "position": {
      "end": {
        "column": 4,
        "line": 1,
      },
      "start": {
        "column": 2,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize positive floats 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "ADDITIVE_OPERATOR",
    "value": "+",
  },
  {
    "position": {
      "end": {
        "column": 6,
        "line": 1,
      },
      "start": {
        "column": 2,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10.5",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize positive integers 1`] = `
[
  {
    "position": {
      "end": {
        "column": 2,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "ADDITIVE_OPERATOR",
    "value": "+",
  },
  {
    "position": {
      "end": {
        "column": 4,
        "line": 1,
      },
      "start": {
        "column": 2,
        "line": 1,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize strings 1`] = `
[
  {
    "position": {
      "end": {
        "column": 14,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "STRING_LITERAL",
    "value": "hello world",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;

exports[`Tokenizer - Literals should tokenize strings with escaped characters 1`] = `
[
  {
    "position": {
      "end": {
        "column": 9,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "STRING_LITERAL",
    "value": "hello ",
  },
  {
    "position": {
      "end": {
        "column": 14,
        "line": 1,
      },
      "start": {
        "column": 9,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "world",
  },
  {
    "position": {
      "end": {
        "column": 16,
        "line": 1,
      },
      "start": {
        "column": 14,
        "line": 1,
      },
    },
    "type": "STRING_LITERAL",
    "value": "",
  },
  {
    "position": {
      "end": {
        "column": -1,
        "line": 1,
      },
      "start": {
        "column": -1,
        "line": 1,
      },
    },
    "type": "END_OF_FILE",
    "value": "EOF",
  },
]
`;
